sweep/sweeper.go

```diff
@@ -29,13 +64,387 @@ type UtxoSweeperConfig struct {
+			if sweepTxHash != *spend.SpenderTxHash {
+				err = ErrRemoteSpend
+
+				// TODO: Add remaining inputs to inputsToSweep.
+			}
+
+			// If remote party is sweeping even before we published
+			// the sweep, we don't need to include it anymore.
+			delete(inputToSweep, *outpoint)
+
+			// Signal result channels sweep result.
+			s.signalResult(outpoint, Result{
+				Tx:  spend.SpendingTx,
+				Err: err,
+			})
+
+		case <-s.cfg.Ticker.Ticks():
+			log.Trace("Tick")
+			// Continuing waiting if nothing to sweep.
+			if len(inputToSweep) == 0 {
+				break
+			}
+
+			// Wait for sweep countdown
+			sweepCountdown--
+			if sweepCountdown > 0 {
+				break
+			}
+
+			var inputList []Input
+			for _, input := range inputToSweep {
+				inputList = append(inputList, input)
+			}
+			sweepTx, err := s.sweep(inputList)
+			if err != nil {
+				log.Error(err)
+
+				// TODO: Rebroadcaster/retry mechanisms need to be added.
+
+				// In case of a double spend, don't notify
+				// listeners yet but wait for the spending tx to
+				// trigger the spend notification.
+				if err == lnwallet.ErrDoubleSpend {
+					clearInputsToSweep()
+					break
+				}
+
+				// In case of an other error, signal all
+				// involved inputs the error.
+				for _, input := range inputToSweep {
+					s.signalResult(input.OutPoint(),
+						Result{Err: err})
+				}
+
+				// Do not try to sweep any of the inputs that
+				// cause this error again.
+				clearInputsToSweep()
+
+				break
+			}
+
+			// Store published transaction hash to detect remote
+			// spend when the spend event comes in.
+			sweepTxHash := sweepTx.TxHash()
+			for _, input := range inputToSweep {
+				outpoint := input.OutPoint()
+				s.pendingInputs[*outpoint].txHash = sweepTxHash
+			}
+
+			// Clear all pending inputs as these have now been
+			// published successfully.
+			clearInputsToSweep()
+		case <-s.quit:
+			return
+		}
+
+		// TODO: Handle the case where a tx is not confirmed within
+		// specified time and fee needs to be bumped.
+	}
+
+}
+
+func (s *UtxoSweeper) signalResult(outpoint *wire.OutPoint, result Result) {
+	pendInput := s.pendingInputs[*outpoint]
+	listeners := pendInput.listeners
+
+	if result.Err == nil {
+		log.Debugf("Dispatching sweep success for %v to %v listeners",
+			outpoint, len(listeners))
+	} else {
+		log.Debugf("Dispatching sweep error for %v to %v listeners: %v",
+			outpoint, len(listeners), result.Err)
+	}
+
+	for _, resultChan := range listeners {
+		s.signalResultChan(resultChan, result)
+	}
+
+	// Cancel spend notification with chain notifier. This is not necessary
+	// in case of a success, except for that a reorg could still happen.
+	pendInput.ntfnRegCancel()
+
+	// Inputs are no longer pending after result has been sent.
+	delete(s.pendingInputs, *outpoint)
+}
+
+func (s *UtxoSweeper) signalResultChan(resultChan chan Result, result Result) {
+	select {
+	case resultChan <- result:
+	case <-s.quit:
+	case <-time.After(sweeperTimeout):
```

Comment:
If we ensure the resultChan has a buffer of 1, then we can remove this extra case, as the send will _always_ succeed.

Reply:
Good point! Changed. I also removed the quit case. Because we only every send a single item, it isn't necessary either?